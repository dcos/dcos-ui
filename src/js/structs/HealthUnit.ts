import Item from "./Item";
import StringUtil from "../utils/StringUtil";
import UnitHealthUtil from "../utils/UnitHealthUtil";

const ID_PREFIX = "dcos-";
const REPLACE_IN_ID = { dns: "DNS" };

class HealthUnit extends Item {
  getHealth() {
    return UnitHealthUtil.getHealth(this.get("health"));
  }

  getTitle() {
    let title = this.get("name");
    if (!title) {
      title = this.getPrettyPrintID();
    }

    return title;
  }

  getPrettyPrintID() {
    let id = this.get("id") || "";
    const prefixIndex = id.indexOf(ID_PREFIX);

    if (prefixIndex >= 0) {
      id = id.slice(prefixIndex + ID_PREFIX.length);
    }

    return StringUtil.idToTitle([id], [".", "-"], REPLACE_IN_ID, true);
  }
}

export default HealthUnit;
