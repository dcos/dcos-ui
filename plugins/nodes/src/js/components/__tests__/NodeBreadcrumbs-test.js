import React from "react";
import renderer from "react-test-renderer";
import HealthUnit from "#SRC/js/structs/HealthUnit";

const mockMesosSummaryStore = {
  getLastSuccessfulSummarySnapshot: jest.fn(),
  addChangeListener: jest.fn()
};
jest.mock("#SRC/js/stores/MesosSummaryStore", function() {
  return mockMesosSummaryStore;
});

const mockUnitHealthStore = {
  getUnit: jest.fn(),
  getNode: jest.fn()
};
jest.mock("#SRC/js/stores/UnitHealthStore", function() {
  return mockUnitHealthStore;
});

const NodeBreadcrumbs = require("../NodeBreadcrumbs");

describe("NodeBreadcrumbs", function() {
  beforeEach(function() {
    mockMesosSummaryStore.getLastSuccessfulSummarySnapshot = jest.fn();
    mockUnitHealthStore.getUnit = jest.fn();
    mockUnitHealthStore.getNode = jest.fn();
  });

  it("renders without breadcrumbs", function() {
    const tree = renderer.create(<NodeBreadcrumbs />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders with node breadcrumbs", function() {
    mockMesosSummaryStore.getLastSuccessfulSummarySnapshot.mockReturnValue({
      slaves: [
        {
          id: "e99adb4a-eee7-4e48-ba86-79cd061d2215-S1",
          hostname: "foo.bar.baz"
        }
      ]
    });
    const tree = renderer
      .create(
        <NodeBreadcrumbs nodeID="e99adb4a-eee7-4e48-ba86-79cd061d2215-S1" />
      )
      .toJSON();

    expect(
      mockMesosSummaryStore.getLastSuccessfulSummarySnapshot
    ).toHaveBeenCalled();
    expect(tree).toMatchSnapshot();
  });

  it("renders with taskID breadcrumbs", function() {
    const tree = renderer
      .create(
        <NodeBreadcrumbs taskID="super-nice-task-id" taskName="My Task" />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders with unitID breadcrumbs", function() {
    mockMesosSummaryStore.getLastSuccessfulSummarySnapshot.mockReturnValue({
      slaves: [
        {
          id: "e99adb4a-eee7-4e48-ba86-79cd061d2215-S1",
          hostname: "foo.bar.baz"
        }
      ]
    });
    mockUnitHealthStore.getUnit.mockReturnValue(
      new HealthUnit({
        title: "MyUnit",
        hostname: "health-unit-hostname"
      })
    );
    mockUnitHealthStore.getNode.mockReturnValue({
      getHealth: jest.fn(function() {
        return {
          classNames: "green",
          title: "All green"
        };
      })
    });
    const tree = renderer
      .create(
        <NodeBreadcrumbs
          nodeID="e99adb4a-eee7-4e48-ba86-79cd061d2215-S1"
          unitID="unit-ids-are-nice"
        />
      )
      .toJSON();

    expect(mockUnitHealthStore.getUnit).toHaveBeenCalled();
    expect(mockUnitHealthStore.getNode).toHaveBeenCalled();

    expect(tree).toMatchSnapshot();
  });

  it("renders no unitID breadcrumbs without nodeID", function() {
    const tree = renderer
      .create(<NodeBreadcrumbs unitID="unit-ids-are-nice" />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders no node breadcrumbs if node cant be found", function() {
    mockMesosSummaryStore.getLastSuccessfulSummarySnapshot.mockReturnValue({
      slaves: []
    });
    const tree = renderer
      .create(
        <NodeBreadcrumbs nodeID="e99adb4a-eee7-4e48-ba86-79cd061d2215-S1" />
      )
      .toJSON();

    expect(
      mockMesosSummaryStore.getLastSuccessfulSummarySnapshot
    ).toHaveBeenCalled();
    expect(tree).toMatchSnapshot();
  });
});
