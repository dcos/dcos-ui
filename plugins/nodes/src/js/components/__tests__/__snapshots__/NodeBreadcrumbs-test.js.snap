// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeBreadcrumbs renders no node breadcrumbs if node cant be found 1`] = `
<div
  className="breadcrumbs"
>
  <div
    className="breadcrumb breadcrumb--is-icon"
  >
    <a
      onClick={[Function]}
      style={Object {}}
    >
      <svg
        aria-label="product-servers icon"
        className="css-zgvqkl"
        height={24}
        preserveAspectRatio="xMinYMin meet"
        role="img"
        viewBox="0 0 24 24"
        width={24}
      >
        <use
          xlinkHref="#product-servers"
        />
      </svg>
    </a>
  </div>
  <div
    className="breadcrumb breadcrumb--is-caret"
  >
    <svg
      aria-label="system-caret-right icon"
      className="css-1vunjz0"
      height={16}
      preserveAspectRatio="xMinYMin meet"
      role="img"
      viewBox="0 0 16 16"
      width={16}
    >
      <use
        xlinkHref="#system-caret-right"
      />
    </svg>
  </div>
  <div
    className="breadcrumb"
  >
    <div
      className="breadcrumb__content breadcrumb__content--text"
    >
      <a
        onClick={[Function]}
        style={Object {}}
      >
        Nodes
      </a>
    </div>
  </div>
</div>
`;

exports[`NodeBreadcrumbs renders no unitID breadcrumbs without nodeID 1`] = `
<div
  className="breadcrumbs"
>
  <div
    className="breadcrumb breadcrumb--is-icon"
  >
    <a
      onClick={[Function]}
      style={Object {}}
    >
      <svg
        aria-label="product-servers icon"
        className="css-zgvqkl"
        height={24}
        preserveAspectRatio="xMinYMin meet"
        role="img"
        viewBox="0 0 24 24"
        width={24}
      >
        <use
          xlinkHref="#product-servers"
        />
      </svg>
    </a>
  </div>
  <div
    className="breadcrumb breadcrumb--is-caret"
  >
    <svg
      aria-label="system-caret-right icon"
      className="css-1vunjz0"
      height={16}
      preserveAspectRatio="xMinYMin meet"
      role="img"
      viewBox="0 0 16 16"
      width={16}
    >
      <use
        xlinkHref="#system-caret-right"
      />
    </svg>
  </div>
  <div
    className="breadcrumb"
  >
    <div
      className="breadcrumb__content breadcrumb__content--text"
    >
      <a
        onClick={[Function]}
        style={Object {}}
      >
        Nodes
      </a>
    </div>
  </div>
</div>
`;

exports[`NodeBreadcrumbs renders with node breadcrumbs 1`] = `
<div
  className="breadcrumbs"
>
  <div
    className="breadcrumb breadcrumb--is-icon"
  >
    <a
      onClick={[Function]}
      style={Object {}}
    >
      <svg
        aria-label="product-servers icon"
        className="css-zgvqkl"
        height={24}
        preserveAspectRatio="xMinYMin meet"
        role="img"
        viewBox="0 0 24 24"
        width={24}
      >
        <use
          xlinkHref="#product-servers"
        />
      </svg>
    </a>
  </div>
  <div
    className="breadcrumb breadcrumb--is-caret"
  >
    <svg
      aria-label="system-caret-right icon"
      className="css-1vunjz0"
      height={16}
      preserveAspectRatio="xMinYMin meet"
      role="img"
      viewBox="0 0 16 16"
      width={16}
    >
      <use
        xlinkHref="#system-caret-right"
      />
    </svg>
  </div>
  <div
    className="breadcrumb"
  >
    <div
      className="breadcrumb__content breadcrumb__content--text"
    >
      <a
        onClick={[Function]}
        style={Object {}}
      >
        Nodes
      </a>
    </div>
  </div>
  <div
    className="breadcrumb breadcrumb--is-caret"
  >
    <svg
      aria-label="system-caret-right icon"
      className="css-1vunjz0"
      height={16}
      preserveAspectRatio="xMinYMin meet"
      role="img"
      viewBox="0 0 16 16"
      width={16}
    >
      <use
        xlinkHref="#system-caret-right"
      />
    </svg>
  </div>
  <div
    className="breadcrumb"
  >
    <div
      className="breadcrumb__content breadcrumb__content--text"
    >
      <a
        onClick={[Function]}
        style={Object {}}
      >
        foo.bar.baz
      </a>
    </div>
  </div>
</div>
`;

exports[`NodeBreadcrumbs renders with taskID breadcrumbs 1`] = `
<div
  className="breadcrumbs"
>
  <div
    className="breadcrumb breadcrumb--is-icon"
  >
    <a
      onClick={[Function]}
      style={Object {}}
    >
      <svg
        aria-label="product-servers icon"
        className="css-zgvqkl"
        height={24}
        preserveAspectRatio="xMinYMin meet"
        role="img"
        viewBox="0 0 24 24"
        width={24}
      >
        <use
          xlinkHref="#product-servers"
        />
      </svg>
    </a>
  </div>
  <div
    className="breadcrumb breadcrumb--is-caret"
  >
    <svg
      aria-label="system-caret-right icon"
      className="css-1vunjz0"
      height={16}
      preserveAspectRatio="xMinYMin meet"
      role="img"
      viewBox="0 0 16 16"
      width={16}
    >
      <use
        xlinkHref="#system-caret-right"
      />
    </svg>
  </div>
  <div
    className="breadcrumb"
  >
    <div
      className="breadcrumb__content breadcrumb__content--text"
    >
      <a
        onClick={[Function]}
        style={Object {}}
      >
        Nodes
      </a>
    </div>
  </div>
  <div
    className="breadcrumb breadcrumb--is-caret"
  >
    <svg
      aria-label="system-caret-right icon"
      className="css-1vunjz0"
      height={16}
      preserveAspectRatio="xMinYMin meet"
      role="img"
      viewBox="0 0 16 16"
      width={16}
    >
      <use
        xlinkHref="#system-caret-right"
      />
    </svg>
  </div>
  <div
    className="breadcrumb"
  >
    <div
      className="breadcrumb__content breadcrumb__content--text"
    >
      <a
        onClick={[Function]}
        style={Object {}}
      >
        My Task
      </a>
    </div>
  </div>
</div>
`;

exports[`NodeBreadcrumbs renders with unitID breadcrumbs 1`] = `
<div
  className="breadcrumbs"
>
  <div
    className="breadcrumb breadcrumb--is-icon"
  >
    <a
      onClick={[Function]}
      style={Object {}}
    >
      <svg
        aria-label="product-servers icon"
        className="css-zgvqkl"
        height={24}
        preserveAspectRatio="xMinYMin meet"
        role="img"
        viewBox="0 0 24 24"
        width={24}
      >
        <use
          xlinkHref="#product-servers"
        />
      </svg>
    </a>
  </div>
  <div
    className="breadcrumb breadcrumb--is-caret"
  >
    <svg
      aria-label="system-caret-right icon"
      className="css-1vunjz0"
      height={16}
      preserveAspectRatio="xMinYMin meet"
      role="img"
      viewBox="0 0 16 16"
      width={16}
    >
      <use
        xlinkHref="#system-caret-right"
      />
    </svg>
  </div>
  <div
    className="breadcrumb"
  >
    <div
      className="breadcrumb__content breadcrumb__content--text"
    >
      <a
        onClick={[Function]}
        style={Object {}}
      >
        Nodes
      </a>
    </div>
  </div>
  <div
    className="breadcrumb breadcrumb--is-caret"
  >
    <svg
      aria-label="system-caret-right icon"
      className="css-1vunjz0"
      height={16}
      preserveAspectRatio="xMinYMin meet"
      role="img"
      viewBox="0 0 16 16"
      width={16}
    >
      <use
        xlinkHref="#system-caret-right"
      />
    </svg>
  </div>
  <div
    className="breadcrumb"
  >
    <div
      className="breadcrumb__content breadcrumb__content--text"
    >
      <a
        onClick={[Function]}
        style={Object {}}
      >
        foo.bar.baz
      </a>
    </div>
  </div>
  <div
    className="breadcrumb breadcrumb--is-caret"
  >
    <svg
      aria-label="system-caret-right icon"
      className="css-1vunjz0"
      height={16}
      preserveAspectRatio="xMinYMin meet"
      role="img"
      viewBox="0 0 16 16"
      width={16}
    >
      <use
        xlinkHref="#system-caret-right"
      />
    </svg>
  </div>
  <div
    className="breadcrumb"
  >
    <div
      className="breadcrumb__content breadcrumb__content--text"
    >
      <a
        onClick={[Function]}
        style={Object {}}
      >
         
        <span
          className="green"
        >
          (
          All green
          )
        </span>
      </a>
    </div>
  </div>
</div>
`;

exports[`NodeBreadcrumbs renders without breadcrumbs 1`] = `
<div
  className="breadcrumbs"
>
  <div
    className="breadcrumb breadcrumb--is-icon"
  >
    <a
      onClick={[Function]}
      style={Object {}}
    >
      <svg
        aria-label="product-servers icon"
        className="css-zgvqkl"
        height={24}
        preserveAspectRatio="xMinYMin meet"
        role="img"
        viewBox="0 0 24 24"
        width={24}
      >
        <use
          xlinkHref="#product-servers"
        />
      </svg>
    </a>
  </div>
  <div
    className="breadcrumb breadcrumb--is-caret"
  >
    <svg
      aria-label="system-caret-right icon"
      className="css-1vunjz0"
      height={16}
      preserveAspectRatio="xMinYMin meet"
      role="img"
      viewBox="0 0 16 16"
      width={16}
    >
      <use
        xlinkHref="#system-caret-right"
      />
    </svg>
  </div>
  <div
    className="breadcrumb"
  >
    <div
      className="breadcrumb__content breadcrumb__content--text"
    >
      <a
        onClick={[Function]}
        style={Object {}}
      >
        Nodes
      </a>
    </div>
  </div>
</div>
`;
